src/test/java/software/amazon/smithy/lsp/language/FoldingRangeHandlerTest.java

```diff
@@ -0,0 +1,463 @@
+                    
+                    }
+                }
+                """);
+
+        List<int[]> ranges = getFoldingRanges(model);
+
+        assertThat(ranges, hasSize(1));
+
+    }
+
+    @Test
+    public void foldingRangeForMultipleAdjacentBlocks() {
+        String model = safeString("""
+                $version: "2"
+                namespace com.foo
+                
+                structure First {
+                    a: String
+                }
+                structure Second {
+                    b: String
+                }
+                structure Third {
+                    c: String
+                }
+                """);
+
+        List<int[]> ranges = getFoldingRanges(model);
+
+        assertThat(ranges, hasSize(3));
+
+    }
+
+    @Test
+    public void foldingRangeForStructureWithComment() {
+        TextWithPositions model = TextWithPositions.from("""
+                $version: "2"
+                namespace com.foo
+                
+                // Comment before
+                structure WithComments {%
+                    // Comment inside
+                    field1: String,
+                    field2: Integer // Inline comment
+                    // Comment between fields
+                    field3: Boolean
+                }% // Comment after
+                """);
+
+        List<int[]> ranges = getFoldingRanges(model.text());
+
+        assertThat(ranges, hasSize(1));
+        assertEquals(ranges.get(0)[0], model.positions()[0].getLine());
+        assertEquals(ranges.get(0)[1], model.positions()[1].getLine());
+    }
+
+
+    @Test
+    public void foldingRangeForNestedStructures() {
+        String model = safeString("""
+                $version: "2"
+                namespace com.foo
+                
+                resource Person {
+                    name: {
+                        firstName: String,
+                        lastName: String
+                    },
+                    address: {
+                        street: String,
+                        city: String,
+                        country: String
+                    }
+                    operations: [GetName,
+                                 GetAddress,
+                                 GetCountry,
+                                 GetStreet,
+                                 GetCity]
+                }
+                """);
+
+        List<int[]> ranges = getFoldingRanges(model);
+
+        assertThat(ranges, hasSize(4));
+        assertArrayEquals(new int[]{3, 18}, ranges.get(0));
+        assertArrayEquals(new int[]{4, 7}, ranges.get(1));
+        assertArrayEquals(new int[]{8, 12}, ranges.get(2));
+        assertArrayEquals(new int[]{13, 17}, ranges.get(3));
+    }
+
+
+
+    @Test
+    public void foldingRangeForMultipleAndNestedTraits() {
+        TextWithPositions model = TextWithPositions.from("""
+                @restJson1%
+                @title("")
+                @cors()
```

Comment:
For cases like this, need to change Parser#traitNode() to not skip the closing ')', since that causes Parser#traitApplication() to eat all the whitespace after the closing ')', setting the end of the TraitApplication to be the start of the next statement